airbyte-workers/src/main/java/io/airbyte/workers/WorkerUtils.java

```diff
@@ -53,17 +56,98 @@ public static void gentleClose(final Process process, final long timeout, final
+   * @param checkHeartbeatTimeUnit - frequency with which the heartbeat of the process is checked.
+   * @param forcedShutdownMagnitude - amount of time to wait if a process needs to be destroyed
+   *        forcibly.
+   * @param forcedShutdownTimeUnit - amount of time to wait if a process needs to be destroyed
+   *        forcibly.
+   * @param forceShutdown
+   */
+  public static void gentleCloseWithHeartbeat(final Process process,
+                                              final HeartbeatMonitor heartbeatMonitor,
+                                              final long gracefulShutdownMagnitude,
+                                              final TimeUnit gracefulShutdownTimeUnit,
+                                              final long checkHeartbeatMagnitude,
+                                              final TimeUnit checkHeartbeatTimeUnit,
+                                              final long forcedShutdownMagnitude,
+                                              final TimeUnit forcedShutdownTimeUnit) {
+    gentleCloseWithHeartbeat(
+        process,
+        heartbeatMonitor,
+        gracefulShutdownMagnitude,
+        gracefulShutdownTimeUnit,
+        checkHeartbeatMagnitude,
+        checkHeartbeatTimeUnit,
+        forcedShutdownMagnitude,
+        forcedShutdownTimeUnit,
+        WorkerUtils::forceShutdown);
+  }
+
+  @VisibleForTesting
+  static void gentleCloseWithHeartbeat(final Process process,
+                                       final HeartbeatMonitor heartbeatMonitor,
+                                       final long gracefulShutdownMagnitude,
+                                       final TimeUnit gracefulShutdownTimeUnit,
+                                       final long checkHeartbeatMagnitude,
+                                       final TimeUnit checkHeartbeatTimeUnit,
+                                       final long forcedShutdownMagnitude,
+                                       final TimeUnit forcedShutdownTimeUnit,
+                                       final TriConsumer<Process, Long, TimeUnit> forceShutdown) {
+
+    while (process.isAlive() && heartbeatMonitor.isBeating()) {
       try {
-        process.waitFor(1, TimeUnit.MINUTES);
+        process.waitFor(checkHeartbeatMagnitude, checkHeartbeatTimeUnit);
       } catch (InterruptedException e) {
-        LOGGER.error("Exception while while killing the process", e);
+        LOGGER.error("Exception while while waiting for process to finish", e);
       }
-      if (process.isAlive()) {
-        LOGGER.warn("Couldn't kill the process. You might have a zombie ({})", process.info().commandLine());
+    }
+
+    if (process.isAlive()) {
+      try {
+        process.waitFor(gracefulShutdownMagnitude, gracefulShutdownTimeUnit);
+      } catch (InterruptedException e) {
+        LOGGER.error("Exception during grace period for process to finish", e);
       }
     }
+
+    // if we have closed gracefully exit...
```

Comment:
```suggestion
    // if we were unable to exit gracefully, force shutdown
```

comment is confusing since the if statement here is the opposite case